src/main/java/co/rsk/federate/signing/hsm/advanceblockchain/HsmBookkeepingClientImpl.java

```diff
@@ -129,35 +154,41 @@ protected void sendBlockHeadersChunks(
-        final String BLOCKS_FIELD = "blocks";
-        ArrayNode blocksFieldData = new ObjectMapper().createArrayNode();
-        for (String blockHeader : blockHeaderChunk) {
-            blocksFieldData.add(blockHeader);
-        }
-        payload.set(BLOCKS_FIELD, blocksFieldData);
-    }
-
-    private void addBrothersToPayload(ObjectNode payload, String[] blockHeaderChunk) {
-        final String BROTHERS_FIELD = "brothers";
-        ArrayNode brothersFieldData = new ObjectMapper().createArrayNode();
-        for (String blockHeader : blockHeaderChunk) {
-            // TODO: This is currently sending empty arrays as brothers to the HSM V3 for compatibility with V2
-            //  This should be changed to sending the actual brothers when HSM V3 is fully implemented
-            brothersFieldData.add(new ObjectMapper().createArrayNode());
-        }
-        payload.set(BROTHERS_FIELD, brothersFieldData);
-    }
-
-    @Override
-    public void updateAncestorBlock(UpdateAncestorBlockMessage updateAncestorBlockMessage) throws HSMClientException {
-        sendBlockHeadersChunks(updateAncestorBlockMessage.getData(), "updateAncestorBlock", true);
-    }
-
     @Override
     public void advanceBlockchain(List<Block> blocks) throws HSMClientException {
+        final String ADVANCE_BLOCKCHAIN_METHOD_NAME = "advanceBlockchain";
         AdvanceBlockchainMessage message = new AdvanceBlockchainMessage(blocks);
-        sendBlockHeadersChunks(message.getParsedBlockHeaders(), "advanceBlockchain", false);
+        List<String> blockHeaders = message.getParsedBlockHeaders();
+        validateBlockHeaders(blockHeaders, ADVANCE_BLOCKCHAIN_METHOD_NAME);
+        List<String[]> blockHeadersChunks = getChunks(blockHeaders.toArray(new String[]{}), maxChunkSize, false);
+
+        logger.trace("[advanceBlockchain] Payload total size: {}", blockHeaders.size());
+        for (int i = 0; i < blockHeadersChunks.size(); i++) {
+            try {
+                String[] blockHeaderChunk = blockHeadersChunks.get(i);
+                ObjectNode payload = this.hsmClientProtocol.buildCommand(ADVANCE_BLOCKCHAIN_METHOD_NAME, getVersion());
+                addBlocksToPayload(payload, blockHeaderChunk);
+
+                if (getVersion() >= 3) {
+                    List<String[]> brothers = getBrothers(blockHeaderChunk, message);
+                    addBrothersToPayload(payload, brothers);
+                }
+
+                if (isStopped) {
+                    return;
+                }
+                logger.trace("[advanceBlockchain] chunk {}/{}", i + 1, blockHeadersChunks.size());
+                this.hsmClientProtocol.send(payload);
+            } catch (HSMClientException e) {
+                logger.warn(
+                    "[advanceBlockchain] failed sending {}/{} chunks. Error: {}",
+                    i + 1,
+                    blockHeadersChunks.size(),
+                    e.getMessage()
+                );
+                throw e;
```

Comment:
Check Sonar alert regarding this. Where is this exception caught? Does it makes sense to catch and re throw?

Reply:
We Log It For Debugging Purpose, Then Throw It To The Caller